### PR TITLE
Add `calculate_shares_delta_given_bonds_delta_derivative`

### PR DESCRIPTION
**Pull request by sentilesdal**
# Resolved Issues
  This is a part of adding functions needed to test remove_liquidity

# Description
  - Adds the lower level function `calculate_shares_delta_given_bonds_delta_derivative` to lp.rs
  - Adds a new version of `calculate_present_value_safe` that doesn't panic.  Future work will remove the previous version.
  - Adds a helper function to calculate idle reserves
  - Tests `calculate_shares_delta_given_bonds_delta_derivative`. 

I ran diagnostics to make sure that we were actually testing the function, the test errors on the present value calculation about half the time, errors on the actual function 1/4 time and succeeds about a 1/4 of the time.

# Review Checklists

Please check each item **before approving** the pull request. While going
through the checklist, it is recommended to leave comments on items that are
referenced in the checklist to make sure that they are reviewed.

- [ ] **Testing**
    - [ ] Are there new or updated unit or integration tests?
    - [ ] Do the tests cover the happy paths?
    - [ ] Do the tests cover the unhappy paths?
    - [ ] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?
    - [ ] If matching Solidity behavior, are there differential fuzz tests that
          ensure that Rust matches Solidity?

### CODE REVIEW

File: crates/hyperdrive-math/src/lp.rs
Comment (jalextowle) [inline crates/hyperdrive-math/src/lp.rs:265]: The use of these non-safe functions means that this isn't "safe."

File: crates/hyperdrive-math/src/lp.rs
Comment (jalextowle) [inline crates/hyperdrive-math/src/lp.rs:486]: Looks like the spacing might be messed up here.

File: crates/hyperdrive-math/src/lp.rs
Comment (jalextowle) [inline crates/hyperdrive-math/src/lp.rs:437]: This is missing the Ascii math from the contract code. There is other Ascii math in the function body that is missing. I think it would be best to just duplicate the comments since the documentation is good.

File: crates/hyperdrive-math/src/lp.rs
Comment (jalextowle) [inline crates/hyperdrive-math/src/lp.rs:866]: This test is amazing. Really nice job dude!

File: crates/hyperdrive-math/src/lp.rs
Comment (jalextowle) [inline crates/hyperdrive-math/src/lp.rs:463]: You're going to want to use a safe variant here.

File: crates/hyperdrive-math/src/lp.rs
Comment (jalextowle) [inline crates/hyperdrive-math/src/lp.rs:474]: Let's just use the same naming as in the contracts, unless there is something that I'm missing. Switching the style is one thing, but I think it'll be easier to follow if it's exactly the same.

File: crates/hyperdrive-math/src/lp.rs
Comment (jalextowle) [inline crates/hyperdrive-math/src/lp.rs:488]: Same here, let's keep the names the same.

File: crates/hyperdrive-math/src/lp.rs
Comment (jalextowle) [inline crates/hyperdrive-math/src/lp.rs:514]: Same here, let's use consistent names.

**Review by jalextowle — APPROVED** [review]
The impl looks right to me. A few style nits that should be addressed before merging, but otherwise LGTM

File: crates/hyperdrive-math/src/lp.rs
Comment (sentilesdal) [reply crates/hyperdrive-math/src/lp.rs:265]: thanks yeah you're right, i was just working on a totally safe version actually.  I'll do that in another PR here shortly.

File: crates/hyperdrive-math/src/lp.rs
Comment (dpaiton) [inline crates/hyperdrive-math/src/lp.rs:419]: I'm confused about this -- `self.long_exposure()` is in units of shares, right? So the variable you're creating is `shares/shares` and is unitless? I think that's fine, it happens, but it's not `long_exposure` anymore right? If so I think the variable name should change to something else.

File: crates/hyperdrive-math/src/lp.rs
Comment (dpaiton) [inline crates/hyperdrive-math/src/lp.rs:424]: I think it's more clear to return in each condition and not assign to a variable. e.g.:
```rs
if self.share_reserves() > long_exposure + self.minimum_share_reserves() {
  self.share_reserves() - long_exposure - self.minimum_share_reserves()
} else {
  fixed!(0)
}
```

File: crates/hyperdrive-math/src/lp.rs
Comment (dpaiton) [inline crates/hyperdrive-math/src/lp.rs:497]: `z_e` should be `z_e(x)`?

File: crates/hyperdrive-math/src/lp.rs
Comment (dpaiton) [inline crates/hyperdrive-math/src/lp.rs:496]: This uses `-t_s` but the code uses `self.time_stretch()`. In solidity we have to map `t_s` to `-t_s` but I think we do not in rust. Honestly I get confused by this so we may want to double check with @jalextowle , but either way I think the comment should match the code in Rust.

Comment (dpaiton) [reply crates/hyperdrive-math/src/lp.rs:496]: Also the code does `c / (...)` but the comment says `c * (...)`

File: crates/hyperdrive-math/src/lp.rs
Comment (dpaiton) [inline crates/hyperdrive-math/src/lp.rs:521]: this is not `k(x)` right? `k` is not dependent on `x`

File: crates/hyperdrive-math/src/lp.rs
Comment (dpaiton) [inline crates/hyperdrive-math/src/lp.rs:441]: can you change `x` to `delta_y` (or `\Delta y` if in latex)? We've been somewhat inconsistent about this elsewhere, I'm trying to move us in the direction of consistently using `x` and `\Delta x` for "base".

Comment (dpaiton) [reply crates/hyperdrive-math/src/lp.rs:441]: assuming I have that right... seeing `y(x)` confuses me further.... Maybe the request is better stated as "can you make sure we use `z` for shares; `y` for bonds; and `x` for base)

File: crates/hyperdrive-math/src/lp.rs
Comment (sentilesdal) [reply crates/hyperdrive-math/src/lp.rs:419]: It must be in base and we're converting to shares.  This was translated from here:

https://github.com/delvtech/hyperdrive/blob/f76f5dacd38664cae6a2539770f30ff595d92141/contracts/src/internal/HyperdriveBase.sol#L575

File: crates/hyperdrive-math/src/lp.rs
Comment (sentilesdal) [reply crates/hyperdrive-math/src/lp.rs:424]: yup

File: crates/hyperdrive-math/src/lp.rs
Comment (sentilesdal) [reply crates/hyperdrive-math/src/lp.rs:497]: The comments are copied from the solidity.  I'd like to keep them the same, if there are things we'd like to change, then matching PRs should go up to fix both here and in solidity.

File: crates/hyperdrive-math/src/lp.rs
Comment (sentilesdal) [reply crates/hyperdrive-math/src/lp.rs:496]: it does match, the -1 part of the exponent is dropped and the base is flipped.  the code is altered from the comment to deal with lack of fixedpoints negative number capability.

File: crates/hyperdrive-math/src/lp.rs
Comment (sentilesdal) [reply crates/hyperdrive-math/src/lp.rs:521]: The comments are copied from the solidity.  I'd like to keep them the same, if there are things we'd like to change, then matching PRs should go up to fix both here and in solidity.

File: crates/hyperdrive-math/src/lp.rs
Comment (sentilesdal) [reply crates/hyperdrive-math/src/lp.rs:441]: same thing here, I wanna keep it exactly the same as the solidity to avoid drift.  love that you're looking at these though (and understanding haha).  can we keep these comment changes outta this PR and then we can sit down and update in both spots?  🙏

File: crates/hyperdrive-math/src/lp.rs
Comment (dpaiton) [reply crates/hyperdrive-math/src/lp.rs:419]: ahh ok, `z = x/c` so maybe just rename to `long_exposure_shares`?

File: crates/hyperdrive-math/src/lp.rs
Comment (dpaiton) [reply crates/hyperdrive-math/src/lp.rs:442]: yup sounds good. We already have issue for it:
https://github.com/delvtech/hyperdrive/issues/867
https://github.com/delvtech/hyperdrive/issues/965

can you link 965 in a TODO comment here?

File: crates/hyperdrive-math/src/lp.rs
Comment (dpaiton) [reply crates/hyperdrive-math/src/lp.rs:497]: Ahh so it's positive in the denominator; got it. Can you add a comment mentioning this? I'm trying to make sure that comments match code as much as possible, or have explanations whenever it doesn't.

File: crates/hyperdrive-math/src/lp.rs
Comment (dpaiton) [reply crates/hyperdrive-math/src/lp.rs:522]: roger; I commented in the issue mentioned above to look into this.

File: crates/hyperdrive-math/src/lp.rs
Comment (dpaiton) [reply crates/hyperdrive-math/src/lp.rs:497]: also fyi there's a general problem that the comment math doesn't always match what is implemented. It's residual from new people adding comments as part of their onboarding and doing it incorrectly :(

**Review by dpaiton — APPROVED** [review]
approved pending replies to our comments

File: crates/hyperdrive-math/src/lp.rs
Comment (sentilesdal) [reply crates/hyperdrive-math/src/lp.rs:265]: I have a PR up for this now, I'll update to use the new version when that PR lands.
https://github.com/delvtech/hyperdrive-rs/actions/runs/8897866606/job/24435373595?pr=11

File: crates/hyperdrive-math/src/lp.rs
Comment (sentilesdal) [reply crates/hyperdrive-math/src/lp.rs:419]: I think the name is right, marking as resolved.

**Pull request merged**
